### PR TITLE
Make errors more human friendly

### PR DESCRIPTION
**Pull request by kfsone**
Per https://github.com/goccmack/gocc/issues/109

The aim of this change is to provide a more human-readable form of error-message in the default (.Error()) version.

Most editors/IDEs recognize the FLC (file, line, column) syntax for errors that typically makes them easy to interact with in tooling.

In particular, a common convention is:

```bnf
FileLineComment
    : Filename ":" LineNo ":" ColumnNo ":" " " "error" ":" " " ErrorMessage
    ;
```

Vim, Emacs and Visual Studio/Code will recognize this as "your build didn't work, and I can tell you why".

The parser knows nothing about filenames, so we can't attempt to print one, the user just needs to prefix printing the error with their filename, e.g.

```
fmt.Printf("%s:%s\n", filename, err.Error())
```

### CODE REVIEW

**Comment by kfsone** [comment]
Stumped as to why the build is failing?

File: example/astx/errors/errors.go [hunk 1]
```diff
@@ -22,35 +27,64 @@ type Error struct {
...
+
 	return w.String()
 }
 
 func (e *Error) Error() string {
 	w := new(strings.Builder)
-	fmt.Fprintf(w, "Error in S%d: %s, %s", e.StackTop, token.TokMap.TokenString(e.ErrorToken), e.ErrorToken.Pos.String())
+
+	// identify the line and column of the error in 'gnu' style so it can be understood
+	// by editors and ides; user will need to rpefix it with a filename.
```
Comment (awalterschulze) [inline example/astx/errors/errors.go:53]:
```suggestion
	// by editors and IDEs; users will need to prefix it with a filename.
```

**Comment by awalterschulze** [comment]
Will review more later sorry

File: example/astx/errors/errors.go [hunk 2]
```diff
@@ -22,35 +27,64 @@ type Error struct {
...
+			// if there's only one expectation, don't describe it in list terms.
+			fmt.Fprint(w, "expected: ", e.ExpectedTokens[0])
+		} else if len(e.ExpectedTokens) == 2 {
+			// for two, use a simple "either" expression
+			fmt.Fprint(w, "expected either ", e.ExpectedTokens[0], " or ", e.ExpectedTokens[1])
+		} else {
+			// for a plurality > 2, use the expression "expected one of" and use an oxford comma to
+			// conclude the list.
+			fmt.Fprintf(w, "expected one of: ")
+			expected := make([]string, len(e.ExpectedTokens))
```
[the remark this answers is not in the document]
Comment (kfsone) [reply example/astx/errors/errors.go:70]: Derped-artifact of iterating before I decided to add the comma; I just put all the tokens into expected and then strings.join()d.  Fixing.

[the remark this answers is not in the document]
Comment (kfsone) [reply example/astx/errors/errors.go]: Resolved in 40525f4ad51965a2a7c46c36aec0f5d0b48ef230.

**Comment by kfsone** [comment]
@awalterschulze I removed the quoting on the e.ErrorToken.Lit, because it often seems to include whitespace: e.g `got: "error\r"` (windows).

**Comment by kfsone** [comment]
> Will review more later sorry

A partial review is fully appreciated :)

**Comment by awalterschulze** [comment]
Sorry for the delay, still trying to find some time to review this.

**Comment by kfsone** [comment]
TIL: The downside to using "git checkout -" all the time is you become a twonk and make pull requests from main/master :)

**Review by awalterschulze — COMMENTED** [review]
Sorry about asking to split things into several steps, but it will just make it much easier to review and talk about.

File: example/astx/errors/errors.go [hunk 3]
```diff
@@ -18,39 +25,66 @@ type Error struct {
...
+		return "expected one of " + strings.Join(tokens, ", ")
+	}
+}
+
 func (e *Error) Error() string {
-	w := new(strings.Builder)
-	fmt.Fprintf(w, "Error in S%d: %s, %s", e.StackTop, token.TokMap.TokenString(e.ErrorToken), e.ErrorToken.Pos.String())
+	// identify the line and column of the error in 'gnu' style so it can be understood
+	// by editors and IDEs; user will need to prefix it with a filename.
+	text := fmt.Sprintf("%d:%d: %s: ", e.ErrorToken.Pos.Line, e.ErrorToken.Pos.Column, Severity)
```
[the remark this answers is not in the document]
Comment (awalterschulze) [reply example/astx/errors/errors.go:72]: This is probably fine to keep. I just need to  see the difference between the two errors being printed out.

File: example/astx/parser/parser.go [hunk 4]
```diff
@@ -114,6 +115,7 @@ func (p *Parser) Error(err error, scanner Scanner) (recovered bool, errorAttrib
 		ErrorToken:     p.nextToken,
 		ErrorSymbols:   p.popNonRecoveryStates(),
 		ExpectedTokens: make([]string, 0, 8),
+		UserContext:    p.UserContext,
```
[the remark this answers is not in the document]
Comment (kfsone) [reply example/astx/parser/parser.go:118]: Bleed over, will remove. My penance for not branching :)

**Comment by kfsone** [comment]
All the user context stuff should now be gone from this branch.

**Comment by kfsone** [comment]
@awalterschulze Apologies for making this so tedious with the rookie merge mistakes, a handful of tiny nuances just enough to throw me off: I'm doing my development across gitlab and bitbucket server, I've gone from being a chair on a bash command-line-golf group to an all out powershell user and I'm still relatively new to the go ecosystem :) I literally could not *see* the Makefile initially, lol.

**Comment by awalterschulze** [comment]
Not a problem, I understand these seemingly silly things can keep one busy for hours.

Problem is now in github the diff shows that almost every file has changed, but when I look at the files, they seem to be the same.
I don't know what went wrong, but this makes reviewing the code really hard.
Do you think it is possible to make something easier to review.
Even it is requires creating a new pull request.

**Comment by kfsone** [comment]
Happy to, gives me the chance to correct my mistake of doing it off master :)

**Comment by kfsone** [comment]
@awalterschulze replaced with https://github.com/goccmack/gocc/pull/115

**Pull request closed without merging**